### PR TITLE
Fix unsubscribe from timezone

### PR DESCRIPTION
**Pull request by klaaspieter**
Unsubscribe from the previous channel if the device's time zone changes.

### CODE REVIEW

File: UnitTests/Tests/Controllers/TRApplicationControllerSpec.m [hunk 1]
```diff
@@ -61,6 +62,44 @@ @interface TRWeatherController (Tests)
...
+            [applicationController subscribeForNotificationsWithDeviceToken:deviceToken];
+
+            OCMVerify([courier unsubscribeFromChannel:initialTimeZone.name]);
+        });
+    });
 });
 
 QuickSpecEnd
+
+void inTimeZone(NSTimeZone *timeZone, void (^callback)(void)) {
```
Comment (sharplet) [inline UnitTests/Tests/Controllers/TRApplicationControllerSpec.m]: Is this being used anywhere?

File: UnitTests/Tests/Controllers/TRApplicationControllerSpec.m [hunk 2]
```diff
@@ -61,6 +62,44 @@ @interface TRWeatherController (Tests)
...
+    describe(@"subscribeForNotificationsWithDeviceToken:", ^{
+        it(@"subscribes to notifications on the channel corresponding to the device's time zone", ^{
+            TRCourierClient *courier = OCMClassMock([TRCourierClient class]);
+            TRApplicationController *applicationController = [TRApplicationController new];
+            applicationController.courier = courier;
+
+            NSData *deviceToken = [NSData new];
+            [applicationController subscribeForNotificationsWithDeviceToken:deviceToken];
+
+            NSString *channel = [[NSTimeZone localTimeZone] name];
```
Comment (sharplet) [inline UnitTests/Tests/Controllers/TRApplicationControllerSpec.m]: What do you think about renaming this variable to `expectedChannel`, and moving its definition into the "arrange" portion of the test before the call to `subscribeForNotificationsWithDeviceToken:`?

File: Tropos/Controllers/TRApplicationController.h [hunk 3]
```diff
@@ -13,4 +13,6 @@
 - (RACSignal *)localWeatherNotification;
 - (void)setMinimumBackgroundFetchIntervalForApplication:(UIApplication *)application;
 
+- (void)subscribeForNotificationsWithDeviceToken:(NSData *)deviceToken;
```
Comment (sharplet) [inline Tropos/Controllers/TRApplicationController.h]: I think this should read `subscribeToNotifications`: generally you subscribe "to" something, not "for" it.

File: Tropos/Controllers/TRApplicationController.m [hunk 4]
```diff
@@ -58,4 +58,20 @@ - (void)setMinimumBackgroundFetchIntervalForApplication:(UIApplication *)applica
...
 
+- (void)subscribeForNotificationsWithDeviceToken:(NSData *)deviceToken
+{
+    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+    NSString *formerChannelKey = @"CourierChannel";
+
+    NSString *channel = [[NSTimeZone localTimeZone] name];
+
+    NSString *formerChannel = [userDefaults stringForKey:formerChannelKey];
+    if (formerChannel && ![channel isEqualToString:formerChannel]) {
```
Comment (sharplet) [inline Tropos/Controllers/TRApplicationController.m]: What do you think about reversing this comparison? That way, if `formerChannel` is `nil` the expression would be a `nil` message-send, and evaluate to false, meaning that you don't need to check for `nil` first.

**Comment by sharplet** [comment]
Looks good 👍 

I assume the problem here is getting notifications for multiple channels? What happens if the subscribe or unsubscribe fails? Is there any way to handle that kind of error at the moment?

**Comment by klaaspieter** [comment]
@sharplet I don't think there is in this 'release' but it's something I just added. You can pass a completion handler that returns a `CourierResult` to the method. The reasoning behind not having it from the start is that it's not super useful in the subscribe case. If a subscribe failed, it'll be retried the next time an app is launched.

File: UnitTests/Tests/Controllers/TRApplicationControllerSpec.m (hunk 1 above)
Comment (klaaspieter) [inline UnitTests/Tests/Controllers/TRApplicationControllerSpec.m]: It is now. I wanted to make sure any time zone changes didn't impact other tests. I'm not sure if it's actually a problem, but I hope this function prevents it either way. Plus it reads kind of nice.

File: Tropos/Controllers/TRApplicationController.m (hunk 4 above)
Comment (klaaspieter) [inline Tropos/Controllers/TRApplicationController.m]: Thanks. It appears I forgot how to Objective-C :)

**Pull request merged**